File: src/net/dinkla/raytracer/hits/Shade.java
package net.dinkla.raytracer.hits;

import net.dinkla.raytracer.materials.Material;
import net.dinkla.raytracer.math.*;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:52:14
 * To change this template use File | Settings | File Templates.
 *
 * Typical lifecycle
 *
 * <pre>
 * Shade sr = world.getAccelerator().hitObjects(world, ray);
 * if (sr.hitsAnObject) {
 *   RGBColor color = sr.material.shade(sr); ...
 * </pre> 
 *
 */
public class Shade extends Hit {

    // for specular highlights, set by Tracer
    public Ray ray;

    // Recursion depth, set by tracer
    public int depth;

    public Shade() {
        super();
        this.depth = 0;
        this.ray = null;
        this.object = null;
    }

    public int getDepth() {
        return depth;
    }

    public Point3D getHitPoint() {
        return ray.linear(t);
    }

    public Point3D getLocalHitPoint() {
        return ray.linear(t);
    }

    public Material getMaterial() {
        return object.getMaterial();
    }

    public Ray getRay() {
        return ray;
    }

    public void setRay(Ray ray) {
        this.ray = ray;
    }

    public void setDepth(int depth) {
        this.depth = depth;
    }

}

File: src/main/kotlin/net/dinkla/raytracer/hits/Shade.kt
package net.dinkla.raytracer.hits

import net.dinkla.raytracer.materials.IMaterial
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.math.Vector3D

class Shade : Hit() {

    // for specular highlights, set by Tracer
    var ray: Ray = Ray(Point3D.ORIGIN, Vector3D.ZERO)

    // Recursion depth, set by tracer
    var depth: Int = 0

    val hitPoint: Point3D
        get() = ray.linear(t)

    val localHitPoint: Point3D
        get() = ray.linear(t)

    val material: IMaterial?
        get() = geometricObject?.material

    init {
        geometricObject = null
    }

    override fun toString(): String {
        return "Shade($ray, $depth, $material)"
    }
}
